bdk_electrum/src/lib.rs

```diff
@@ -0,0 +1,800 @@
+                break update;
+            }
+        };
+
+        let last_active_index = request_spks
+            .into_keys()
+            .filter_map(|k| {
+                scanned_spks
+                    .range((k.clone(), u32::MIN)..=(k.clone(), u32::MAX))
+                    .rev()
+                    .find(|(_, (_, active))| *active)
+                    .map(|((_, i), _)| (k, *i))
+            })
+            .collect::<BTreeMap<_, _>>();
+
+        Ok(ElectrumUpdate {
+            update,
+            last_active_indices: last_active_index,
+        })
+    }
+
+    /// Convenience method to call [`scan`] without requiring a keychain.
+    ///
+    /// [`scan`]: ElectrumClient::scan
+    pub fn scan_without_keychain<P, S>(
+        &self,
+        to_position: impl Fn(ChainPositionArgs<'_>) -> Result<P, Error>,
+        local_chain: &BTreeMap<u32, BlockHash>,
+        params: ScanParamsWithoutKeychain<S>,
+    ) -> Result<ElectrumUpdate<(), P>, Error>
+    where
+        P: ChainPosition,
+        S: IntoIterator<Item = Script>,
+    {
+        let spk_iter = params
+            .spks
+            .into_iter()
+            .enumerate()
+            .map(|(i, spk)| (i as u32, spk));
+        let params = ScanParams {
+            keychain_spks: [((), spk_iter)].into(),
+            txids: params.txids,
+            txs: params.txs,
+            outpoints: params.outpoints,
+            stop_gap: params.stop_gap,
+            batch_size: params.batch_size,
+        };
+        self.scan(to_position, local_chain, params)
+    }
+}
+
+/// Parameters for [`ElectrumClient::scan`].
+pub struct ScanParams<K, S: IntoIterator<Item = (u32, Script)>> {
+    /// Indexed script pubkeys of each keychain to scan transaction histories for.
+    pub keychain_spks: BTreeMap<K, S>,
+    /// Txids to scan for. The update will update the [`ChainPosition`]s for these.
+    pub txids: Vec<Txid>,
+    /// Full transactions to scan for. The update will update the [`ChainPosition`]s for these.
+    pub txs: Vec<Transaction>,
+    /// Outpoints to scan for. The update will try include the transaction that spends this outpoint
+    /// alongside the transaction which contains this outpoint.
+    pub outpoints: Vec<OutPoint>,
+    /// The theshold number of [`ScanParams::keychain_spks`] that return empty histories before we
+    /// stop scanning for `keychain_spks`.
+    pub stop_gap: usize,
+    /// The batch size to use for requests that can be batched.
+    pub batch_size: usize,
+}
+
+impl<K, S: IntoIterator<Item = (u32, Script)>> Default for ScanParams<K, S> {
+    fn default() -> Self {
+        Self {
+            keychain_spks: Default::default(),
+            txids: Default::default(),
+            txs: Default::default(),
+            outpoints: Default::default(),
+            stop_gap: 10,
+            batch_size: 10,
+        }
+    }
+}
+
+impl<K, S: IntoIterator<Item = (u32, Script)>> From<BTreeMap<K, S>> for ScanParams<K, S> {
```

Comment:
Make a constructor for this I think. Not sure what `From` is getting us since it makes it harder for the user to discover it.

Reply:
Actually, this is not needed. Removed.